### ISSUE TITLE
[Ray custom environment : A2C take long time to start training and increase temporary file size and ray results directory ] 

### ISSUE THREAD

**Issue by abdjea** [ray-project/ray#27368]
### What happened + What you expected to happen

When I'm launching A2C agent training over my custom environment, it causes an increasing of generated data during the training process in these following files /ray/results and more specifically in /tmp/ray. 

My VM configuration has 30 CPU, 3 GPU, 47 GB RAM and 80 GB space in disk. 

The first observation was as following : 
- the training take too much time to start when I'm trying to use the full data ( 1.76 GB) in my custom environment. It is blocked to dashboard message for a while (View the Ray dashboard at (http://127.0.0.1:8265) .
- When the training is started Mentioned directory started to be filled until no memory space error occurred. I'm using the following config setup : 

ray.init(
    #local_dir="/media/abhishek/HD", not existing
    #_temp_dir="/media/abhishek/HD", not working
    #_plasma_directory="/media/abhishek/HD", not working
    _redis_max_memory=int(1e8),
    object_store_memory=4294967296,
    _system_config={
        "object_spilling_config": json.dumps(
            {"type": "filesystem", "params": {"directory_path": "/media/abhishek/HD"}},
        )
    },
)

config={"env":VideoEnv,
        "env_config": {"data": data_train[:64]},
        "num_workers":0,
        "remote_worker_envs": True,
        #'num_envs_per_worker': 1,
        "num_gpus":3,
        #"horizon":32,
        "soft_horizon":True,
        "no_done_at_end":False,
        "num_envs_per_worker": 2,
       }

- When I'm trying to change the temporary file and results location to my external SSD, it's not reponding.

The second observation was when I switch from 0 to 1 worker : 
- the ray actor is no more responding.

You can see below the custom environment code : It's an environment developed for an RL agent to deal with a binary classification task.

class VideoEnv(gym.Env):

    def __init__(self, config: EnvContext):
        super(VideoEnv, self).__init__()
        self.action_space = spaces.Discrete(2) 
        self.index = 0
        self.data=config["data"]
        self.X = self.data[:,:4096]
        self.Y = self.data[:,-1:]
        
        self.observation_space = spaces.Box(low=-1, high=1, shape=(1, self.X.shape[1]), dtype=np.float32)

    def step(self, action):
        reward = 1 if action == self.Y[self.index] else -1 #ERROR
        done = self.index >= len(self.X)- 1
        if not done:
            self.index += 1

            obs = self.X[self.index]
        elif done:
            obs = self.X[self.index-1] # dummy obs bcz no next state after the last one 

        return np.array([obs]).astype(np.float32), reward, done, {} # action, self.Y[self.index - 1]

    def reset(self):
        self.index = 0
         obs=self.X[self.index]
        return np.array([obs]).astype(np.float32)

### Versions / Dependencies

Ray 1.13
Python 3.8
Ubuntu 18.04
Tensorflow-gpu 2.4
Gym 0.21

### Reproduction script

import gym
from ray.rllib.agents import a3c
import ray
from ray.rllib.agents.a3c import A2CTrainer, DEFAULT_CONFIG
from ray.tune.registry import register_env
from ray import tune
import ray.util
import json

if ray.is_initialized():
    ray.shutdown()

ray.init(
    #local_dir="/media/abhishek/HD", not existing
    #_temp_dir="/media/abhishek/HD", not working
    #_plasma_directory="/media/abhishek/HD", not working
    _redis_max_memory=int(1e8),
    object_store_memory=4294967296,
    _system_config={
        "object_spilling_config": json.dumps(
            {"type": "filesystem", "params": {"directory_path": "/media/abhishek/HD"}},
        )
    },
)

config={"env":VideoEnv,
        "env_config": {"data": data_train},
        "num_workers":2,
        "remote_worker_envs": True,
        #'num_envs_per_worker': 1,
        "num_gpus":3,
        #"horizon":32,
        "soft_horizon":True,
        "no_done_at_end":False,
        #"num_envs_per_worker": 2,
        #"framework": "torch",
       } #,"rollout_fragment_length": 10,"batch_mode": "complete_episodes",}

trainer = a3c.A2CTrainer(config=config) 

for i in range(500):
    print(trainer.train()["episode_reward_mean"])
    trainer.save("/media/abhishek/HD/rllib_checkpoint")

### Issue Severity

High: It blocks me from completing my task.

**Comment by kouroshHakha** [comment]
Hi @abdjea Can you please provide a minimal repro script so that we can test on our end (something that does not include personal drives)? Please also put the code into three quotes for proper markdown formatting. Does this issue happen only when you use your custom environment? Have you tried some other simpler env like cartpole?

**Comment by abdjea** [comment]
Hi @kouroshHakha , thank you for your help. I'm preparing the repro. This happen only with my personal environment and not with cartpole.

**Comment by abdjea** [comment]
Hi @kouroshHakha , you have the script with simulated data just below without personal drive paths : 

"""
import numpy as np

#Data simulation
nbr_dim=4096
nbr_seg=46304

list_seg = np.random.rand(nbr_seg,nbr_dim,).astype(np.float32)
list_label = np.random.randint(0,2,(nbr_seg,))

print(list_seg.shape)
print(list_label.shape)

#Create custom environment
import gym
from gym import spaces
import ray
from ray.rllib.env.env_context import EnvContext

class VideoEnv(gym.Env):

    def __init__(self, config: EnvContext):
        super(VideoEnv, self).__init__()
        # self.env = gym.make(choose_env_for(env_config.worker_index, env_config.vector_index))
        self.action_space = spaces.Discrete(2) 
        # print(X, X.shape)
        self.index = 0
        self.X = config["X"]
        self.Y = config["Y"]
        
        self.observation_space = spaces.Box(low=-1, high=1, shape=(1, self.X.shape[1]), dtype=np.float32)

        # self.num=0

    def step(self, action):
        # print("act env",action)
        # print("label1 : ",self.Y[self.index])
        reward = 1 if action == self.Y[self.index] else -1 #ERROR
        done = self.index >= len(self.X)- 1
        # print(done) #debug when environement done with a dataset
        if not done:
            self.index += 1

            obs = self.X[self.index]
        elif done:
            obs = self.X[self.index-1] # dummy obs bcz no next state after the last one 
            # print("label2 : ",self.Y[self.index - 1])

        return np.array([obs]).astype(np.float32), reward, done, {} # action, self.Y[self.index - 1]

    def reset(self):
        # self.num=1
        self.index = 0
        obs=self.X[self.index]
        return np.array([obs]).astype(np.float32)

#Train ray agent A3C
import ray
from ray.rllib.agents.a3c import A2CTrainer, DEFAULT_CONFIG

import gym
from ray.rllib.agents import a3c
import ray
from ray.rllib.agents import ppo
from ray.tune.registry import register_env
import grpc
from ray import tune
import ray.util
import json

if ray.is_initialized():
    ray.shutdown()

ray.init()

config={"env":VideoEnv,
        "env_config": {"X": list_seg,"Y":list_label},
        "num_workers":0, #Not working when > 0.
        "remote_worker_envs": True,
        #'num_envs_per_worker': 1,
        "num_gpus":3,
        #"horizon":32,
        "soft_horizon":True,
        "no_done_at_end":False,
        "num_envs_per_worker": 2,
       }


trainer = a3c.A2CTrainer(config=config) 


for i in range(500):
    print(trainer.train()["episode_reward_mean"])


"""

This script is not working when the num_workers > 0. My goal is to parallelize environments in order to speed up the training time.
